packages/grid/_modules_/grid/hooks/features/export/useCsvExport.tsx

```diff
@@ -0,0 +1,65 @@
+import * as React from 'react';
+import { ApiRef } from '../../../models/api/apiRef';
+import { useApiMethod } from '../../root/useApiMethod';
+import { useGridSelector } from '../core/useGridSelector';
+import { visibleColumnsSelector } from '../columns';
+import { visibleSortedRowsSelector } from '../filter';
+import { selectionStateSelector } from '../selection';
+import { Columns, CsvExportApi, RowId, RowModel } from '../../../models';
+import { useLogger } from '../../utils/useLogger';
+import { exportAs } from '../../../utils';
+
+const buildRow = (row: RowModel, columns: Columns): Array<RowModel> => {
+  const mappedRow: RowModel[] = [];
+  columns.forEach((column) => column.field !== '__check__' && mappedRow.push(row[column.field]));
+  return mappedRow;
+};
+
+const buildCSV = (
+  columns: Columns,
+  rows: RowModel[],
+  selectedRows: Record<RowId, boolean>,
+): string => {
+  const selectedRowsIds = Object.keys(selectedRows);
+
+  if (selectedRowsIds.length) {
+    rows = rows.filter((row) => selectedRowsIds.includes(`${row.id}`));
+  }
+
+  const CSVHead = `${columns
+    .filter((column) => column.field !== '__check__')
+    .map((column) => column.headerName)
+    .toString()}\r\n`;
+  const CSVBody = rows.reduce((soFar, row) => `${soFar}${buildRow(row, columns)}\r\n`, '').trim();
+  const csv = `${CSVHead}${CSVBody}`.trim();
+
+  return csv;
+};
+
+export const useCsvExport = (apiRef: ApiRef): void => {
+  const logger = useLogger('useCsvExport');
+  const visibleColumns = useGridSelector(apiRef, visibleColumnsSelector);
+  const visibleSortedRows = useGridSelector(apiRef, visibleSortedRowsSelector);
+  const selection = useGridSelector(apiRef, selectionStateSelector);
+
+  const getDataAsCsv = React.useCallback((): string => {
+    logger.debug(`Get data as CSV`);
+
+    return buildCSV(visibleColumns, visibleSortedRows, selection);
+  }, [logger, visibleColumns, visibleSortedRows, selection]);
+
+  const exportDataAsCsv = React.useCallback((): void => {
+    logger.debug(`Export data as CSV`);
+    const csv = getDataAsCsv();
+    const blob = new Blob([csv], { type: 'text/csv' });
+
+    exportAs(blob, 'csv', 'data');
```

Comment:
We could abstract all that in a method like 

```tsx
// csvUtils.ts
function buildRow()...
function serialiseCSV(rows, cols) {
...
   const blob = new Blob([csv], { type: 'text/csv' });

    exportAs(blob, 'csv', 'data');
}

logger.debug(`Export data as CSV`);
const rowsToExport = selectedRows.length > 0 ? selectedRows.map(...) : rows;
exportAsCSV(columns, rowsToExport)

```

Reply:
I could, my only argument against it is are we going to use it somewhere else? If this is the only place where that util is going to be used then it isn't really a utility function. The benefit of the current setup is that the logic is easier to follow.

Reply:
The idea is to abstract the logic between the csv serialization and the setup of the feature in the grid. By splitting the file, you split the responsibility and you'll have a cleaner architecture with a more readable hook. If there is an issue with the csv, you'll know it will be in the serializer...

You could also manage different serializer at a later stage...

Reply:
I don't really have a preference. Personally, I would opt to keep it contained since it is a relatively small chunk of logic that is not meant to be used anywhere else. Looking into the future what I can do is move it to generic export serializers file that will also hold the logic for PDF export in the future if you are ok with this @dtassone

Reply:
@dtassone updated